RFC: opal_clock_gettime() and opal_clock_getres()

jsquyres opened the pull request:
This is an RFC for simple portable wrappers around clock_gettime(3) and clock_getres(3) (and falling back to gettimeofday(3) if clock_gettime is not available).

This PR came out of the discussion [here](https://github.com/open-mpi/ompi/pull/8057#discussion_r762612710) and [here](https://github.com/open-mpi/ompi/pull/8057#discussion_r762618783) (which have additional links to other and older conversations).

If this RFC is accepted, #8057 should be updated to use `opal_clock_gettime()` and `opal_clock_getres()`.

jsquyres commented:
bot:ompi:retest

jsquyres commented:
bot:ompi:retest

bwbarrett commented on ompi/mpi/c/wtime.c:
```diff
@@ -65,36 +58,15 @@ double MPI_Wtime(void)
-    (void) clock_gettime(CLOCK_MONOTONIC, &tp);
-    if( OPAL_UNLIKELY(0 == ompi_wtime_time_origin.tv_sec) ) {
+    (void) opal_clock_gettime(&tp);
+    if (OPAL_UNLIKELY(0 == ompi_wtime_time_origin.tv_sec)) {
```
Do you remember why we bother doing this work?

jsquyres replied:
I do not.  I seem to remember @bosilca having opinions about this (i.e., that we return timestamps relative to the first time Wtime is invoked).  George: do you remember?

jsquyres replied:
I talked to @bosilca about this verbally today.  His recollection is that the tools community asked for this, so that Wtime values would be smaller and therefore potentially compressible (i.e., have multiple bytes that are zero).

bwbarrett reviewed (approved):
Looks good to me.

jsquyres commented:
@bwbarrett I changed the first commit:

* It struck me that if we do not have `clock_gettime()`, we might not have `struct timespec` as well.  So I added configury for that, and made that portable.
* There was a bug in `opal_clock_getres()` in using gettimeofday to set usec precsion.  Fixed.

jsquyres commented:
I talked this PR through with @bosilca verbally today, and he's ok with it.  Merging.
